libs/contrib/Control/Algebra/ComplexImplementations.idr

```diff
@@ -0,0 +1,219 @@
+module Control.Algebra.ComplexImplementations
+
+import Data.Complex
+import Control.Algebra
+import Control.Algebra.Laws
+import Control.Algebra.VectorSpace
+
+%access public export
+%default total
+
+Semigroup a => Semigroup (Complex a) where
+  (<+>) (a :+ b) (c :+ d) = (a <+> c) :+ (b <+> d)
+
+Monoid a => Monoid (Complex a) where
+  neutral = (neutral :+ neutral)
+
+Group a => Group (Complex a) where
+  inverse (r :+ i) = (inverse r :+ inverse i)
+
+Ring a => Ring (Complex a) where
+  (<.>) (a :+ b) (c :+ d) = (a <.> c <-> b <.> d) :+ (a <.> d <+> b <.> c)
+
+RingWithUnity a => RingWithUnity (Complex a) where
+  unity = (unity :+ neutral)
+
+RingWithUnity a => Module a (Complex a) where
+  (<#>) x = map (x <.>)
+
+RingWithUnity a => InnerProductSpace a (Complex a) where
+  (x :+ y) <||> z = realPart $ (x :+ inverse y) <.> z
+
+----------------------------------------
+
+VerifiedSemigroup a => VerifiedSemigroup (Complex a) where
+  semigroupOpIsAssociative (a :+ x) (b :+ y) (c :+ z) =
+    rewrite semigroupOpIsAssociative a b c in
+      rewrite semigroupOpIsAssociative x y z in
+        Refl
+
+VerifiedMonoid t => VerifiedMonoid (Complex t) where
+  monoidNeutralIsNeutralL (a :+ x) =
+    rewrite monoidNeutralIsNeutralL a in
+      rewrite monoidNeutralIsNeutralL x in
+        Refl
+
+  monoidNeutralIsNeutralR (a :+ x) =
+    rewrite monoidNeutralIsNeutralR a in
+      rewrite monoidNeutralIsNeutralR x in
+        Refl
+
+VerifiedGroup t => VerifiedGroup (Complex t) where
+  groupInverseIsInverseR (r :+ i) =
+    rewrite groupInverseIsInverseR r in
+      rewrite groupInverseIsInverseR i in
+        Refl
+
+AbelianGroup t => AbelianGroup (Complex t) where
+  abelianGroupOpIsCommutative (a :+ i) (b :+ j) =
+    rewrite abelianGroupOpIsCommutative a b in
+      rewrite abelianGroupOpIsCommutative i j in
+        Refl
+
+-- A simple helper lemma
+private abelianGroupRearrange : AbelianGroup t => (a, b, c, d : t) ->
+  a <+> b <+> (c <+> d) = a <+> c <+> (b <+> d)
+abelianGroupRearrange a b c d =
+  rewrite sym $ semigroupOpIsAssociative a b (c <+> d) in
+    rewrite semigroupOpIsAssociative b c d in
+      rewrite abelianGroupOpIsCommutative b c in
+    rewrite sym $ semigroupOpIsAssociative c b d in
+  semigroupOpIsAssociative a c (b <+> d)
```

Comment:
It would be nice to have something like that, but I don't know how. I've just been grinding through on a case-by-case basis.

Reply:
I imagine it would be a type containing a `List` (or `Vect`) of terms, with a proof that folding it is the same as a regular use of the function you fold with. It should only require associativity I guess? Then you can implement transformations on the List, proving they still don't change the result (this is where commutativity will be required. I have tried something like this:

```
data Sum : Type -> Type where
    MkSum : Monoid a => List a -> Sum a

eval : Sum a -> a
terms : Sum a -> List a

evalEquvalentToFold : Monoid s => (s : Sum a) -> eval s = fold (<+>) Algebra.neutral  (terms s)
```

Unfortunately proving anything about `fold` is difficult due to the way it's implemented. Probably a slightly different approach would be required, but I think this is the right direction.